src/main/src/main/java/org/geoserver/security/SecuredResourceNameChangeListener.java

```diff
@@ -0,0 +1,170 @@
+
+    static final Logger LOGGER = Logging.getLogger(SecuredResourceNameChangeListener.class);
+
+    Catalog catalog;
+    DataAccessRuleDAO dao;
+
+    public SecuredResourceNameChangeListener(Catalog catalog, DataAccessRuleDAO dao) {
+        this.catalog = catalog;
+        this.dao = dao;
+        catalog.addListener(this);
+    }
+
+    @Override
+    public void handleAddEvent(CatalogAddEvent event) throws CatalogException {
+        // ignore
+    }
+
+    @Override
+    public void handleRemoveEvent(CatalogRemoveEvent event) throws CatalogException {
+        // ignore
+    }
+
+    @Override
+    public void handleModifyEvent(CatalogModifyEvent event) throws CatalogException {
+        // ignore
+    }
+
+    @Override
+    public void handlePostModifyEvent(CatalogPostModifyEvent event) throws CatalogException {
+        // no names where change
+        if (event.getPropertyNames().indexOf("name") == -1) return;
+        // if workspace has been renamed
+        if (event.getSource() instanceof WorkspaceInfo) {
+
+            String oldWorkspaceName =
+                    String.valueOf(
+                            event.getOldValues().get(event.getPropertyNames().indexOf("name")));
+            String newWorkspaceName =
+                    String.valueOf(
+                            event.getNewValues().get(event.getPropertyNames().indexOf("name")));
+            // dont go anything if name has not changed
+            if (oldWorkspaceName.equalsIgnoreCase(newWorkspaceName)) return;
+
+            // if a security rule exists for the workspaec whose name has been changed
+            if (workspaceHasSecurityRule(oldWorkspaceName)) {
+                LOGGER.info(
+                        String.format(
+                                "Updating Security Rules for Renamed Workspace: %s",
+                                oldWorkspaceName));
+
+                // get rules in which workspace name should be updated
+                List<DataAccessRule> rulesToModifyList =
+                        getDataAccessRule(event.getSource(), oldWorkspaceName);
+                for (DataAccessRule rule : rulesToModifyList) rule.setRoot(newWorkspaceName);
+
+                try {
+                    dao.storeRules();
+                } catch (IOException e) {
+                    LOGGER.log(Level.SEVERE, e.getMessage(), e);
+                }
+            }
+        } else if (event.getSource() instanceof ResourceInfo) {
+            // if a layer has been renamed
+            String oldResourceName =
+                    String.valueOf(
+                            event.getOldValues().get(event.getPropertyNames().indexOf("name")));
+            String newResourceName =
+                    String.valueOf(
+                            event.getNewValues().get(event.getPropertyNames().indexOf("name")));
+            // dont do anything if name has not changed
+            if (oldResourceName.equalsIgnoreCase(newResourceName)) return;
```

Comment:
The logic up to here is the same as that for workspace - you could consolidate and move this part outside of the if statements (ultimately not important, and up to personal preference, but that approach would be a bit more concise)